### PR TITLE
Change lager to logger

### PR DESCRIPTION
**Pull request by terry-xiaoyu**
I'd like to optimize the `Log` section in our configuration file, mainly for readability.

### CODE REVIEW

File: etc/emqx.conf [hunk 1]
```diff
@@ -328,83 +328,61 @@ rpc.socket_keepalive_count = 9
...
-log.console = console
+## Value: off | file | stdio | both
+## - off: disable logs entirely
+## - file: write logs to file
+## - stdio: write logs to standard I/O
+## - both: write logs both to file and standard I/O
+log.to = stdio
 
-## Sets the severity level of console log.
+## Sets the severity level for logs that written to stdio.
```
Comment (spring2maz) [inline etc/emqx.conf]:
```suggestion
## Sets the severity level for logs written to stdio.
```

File: etc/emqx.conf [hunk 2]
```diff
@@ -328,83 +328,61 @@ rpc.socket_keepalive_count = 9
...
 ##
 ## Value: debug | info | notice | warning | error | critical | alert | emergency
 ##
 ## Default: error
-log.console.level = error
+log.stdio.level = error
 
-## The file where console logs will be writed to, when 'log.console' is set as 'file'.
+## Sets the severity level for logs that written to file.
```
Comment (spring2maz) [inline etc/emqx.conf]:
```suggestion
## Sets the severity level for logs written to file.
```

**Comment by terry-xiaoyu** [comment]
Other changes will come soon. I created the PR early to discuss the config template with you.

File: etc/emqx.conf [hunk 3]
```diff
@@ -328,83 +328,61 @@ rpc.socket_keepalive_count = 9
...
-##
-## Value: File Name
-log.crash.file = {{ platform_log_dir }}/crash.log
+## Format: log.file.$level.max_num = $number
+##    where "$level" can be one of:
+##    debug, info, notice, warning, error, critical, alert, emergency
+log.file.error.max_num = 5
 
-## Enable syslog.
+## Enable syslog, this will write logs to the rsyslog
```
Comment (spring2maz) [inline etc/emqx.conf]: if we are to deprecate lager, we should delete syslog ?

Comment (spring2maz) [inline etc/emqx.conf]:
```suggestion
## Enable syslog, this will write logs to the syslog
```

File: etc/emqx.conf [hunk 4]
```diff
@@ -328,83 +328,58 @@ rpc.socket_keepalive_count = 9
...
-
-## Enable syslog.
+## Format: log.$level.file = $filename,
+##         where "$level" can be one of: debug, info, notice, warning,
+##                                       error, critical, alert, emergency
+## Note: Log files for a specific log level will contain all the logs
+##       that greater than or equal to that level
+log.error.file = error.log
+
+## Enable syslog, to write logs to the rsyslog
```
Comment (spring2maz) [inline etc/emqx.conf]:
```suggestion
## Enable syslog, to write logs to the syslog
```

File: etc/emqx.conf (hunk 3 above)
Comment (terry-xiaoyu) [reply etc/emqx.conf]: deleted

Comment (terry-xiaoyu) [reply etc/emqx.conf]: deleted

File: etc/emqx.conf [hunk 5]
```diff
@@ -328,91 +328,61 @@ rpc.socket_keepalive_count = 9
...
-log.syslog = on
-
-## Sets the severity level for syslog.
-##
-## Value: debug | info | notice | warning | error | critical | alert | emergency
-log.syslog.level = error
+## Format: log.$level.file = $filename,
+##         where "$level" can be one of: debug, info, notice, warning,
+##                                       error, critical, alert, emergency
+## Note: Log files for a specific log level will contain all the logs
```
Comment (spring2maz) [inline etc/emqx.conf]: I'm not quite sure why we need this extra log config if we already have `log.file`.
Especially when the severity level filter logic (higher than or equal to) is exactly the same.

File: src/emqx_tracer.erl [hunk 6]
```diff
@@ -73,37 +81,45 @@ lookup_traces() ->
...
 handle_call({start_trace, Who, LogFile}, _From, State = #state{level = Level, traces = Traces}) ->
-    case catch lager:trace_file(LogFile, [Who], Level, ?OPTIONS) of
-        {ok, exists} ->
-            {reply, {error, already_exists}, State};
-        {ok, Trace} ->
-            {reply, ok, State#state{traces = maps:put(Who, {Trace, LogFile}, Traces)}};
+    case logger:add_handler(handler_id(Who), logger_disk_log_h,
+                                #{level => Level,
+                                  formatter => ?FORMAT,
+                                  filesync_repeat_interval => 1000,
```
Comment (spring2maz) [inline src/emqx_tracer.erl]: same here fsync interval 1000 is too aggressive IMO

File: src/emqx_logger_formatter.erl [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+    A = do_format(Level,Meta,AT,Config),
+    MsgStr =
+        if DoMsg ->
+                Config1 =
+                    case maps:get(chars_limit,Config) of
+                        unlimited ->
+                            Config;
+                        Size0 ->
+                            Size =
+                                case Size0 - string:length([B,A]) of
```
Comment (spring2maz) [inline src/emqx_logger_formatter.erl]: `string:length([B,A])` is always `2` no matter what `B` and `A` are.
```
1> string:len([a,b]).
2
```

File: src/emqx_logger_formatter.erl [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+    {BT,AT0} = lists:splitwith(fun(msg) -> false; (_) -> true end, Template),
+    {DoMsg,AT} =
+        case AT0 of
+            [msg|Rest] -> {true,Rest};
+            _ ->{false,AT0}
+        end,
+    B = do_format(Level,Meta,BT,Config),
+    A = do_format(Level,Meta,AT,Config),
+    MsgStr =
+        if DoMsg ->
```
Comment (spring2maz) [inline src/emqx_logger_formatter.erl]: this whole block from this line to line 80 worth an abstraction to a help function.

Comment (terry-xiaoyu) [reply src/emqx_logger_formatter.erl]: Agree. But this whole file is copied from OTP `lib/kernel/src/logger_formatter.erl` maily for new date format, and I'd like to keep most of the file untouched.

File: src/emqx_logger_formatter.erl (hunk 7 above)
Comment (terry-xiaoyu) [reply src/emqx_logger_formatter.erl]: `string:length(["B1","A"])` returns 3, I think this code is correct.

**Pull request merged**